src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,263 @@
+                .SelectMany(x => x.Value.diagnostics, (k, v) => ((k.Value.name, v)))
+                .ToImmutableArray();
+        }
+
+        public void QueueForAnalysis(ImmutableArray<ProjectId> projects)
+        {
+            foreach (var projectId in projects)
             {
-                _queueRunning = true;
+                _workQueue.AddOrUpdate(projectId,
+                    (modified: DateTime.UtcNow, projectId: projectId, new CancellationTokenSource()),
+                    (_, oldValue) => (modified: DateTime.UtcNow, projectId: projectId, oldValue.workReadySource));
             }
+        }
 
-            Task.Factory.StartNew(async () =>
-            {
-                await Task.Delay(100);
-                await Dequeue();
+        private ImmutableArray<Task> WaitForPendingWorkIfNeededAndGetIt(ImmutableArray<ProjectId> projectIds)
+        {
+            return _workQueue
+                .Where(x => projectIds.Any(pid => pid == x.Key))
+                .Select(x => Task.Delay(30 * 1000, x.Value.workReadySource.Token)
+                    .ContinueWith(task => LogTimeouts(task, x.Key.ToString())))
+                .Concat(new[] { Task.Delay(250) }) // Workaround for issue where information about updates from workspace are not at sync with calls.
+                .ToImmutableArray();
+        }
 
-                if (_openDocuments.IsEmpty)
-                {
-                    lock (_lock)
-                    {
-                        _queueRunning = false;
-                    }
-                }
-                else
-                {
-                    this.ProcessQueue();
-                }
-            });
+        // Editors seems to fetch initial (get all) diagnostics too soon from api,
+        // and when this happens initially api returns nothing. This causes nothing
+        // to show until user action causes editor to re-fetch all diagnostics from api again.
+        // For this reason initially api waits for results for moment. This isn't perfect
+        // solution but hopefully works until event based diagnostics are published.
+        private Task WaitForInitialStartupWorkIfAny()
+        {
+            return Task.Delay(30 * 1000, _initializationQueueDoneSource.Token)
+                        .ContinueWith(task => LogTimeouts(task, nameof(_initializationQueueDoneSource)));
+        }
+
+        // This is basically asserting mechanism for hanging analysis if any. If this doesn't exist tracking
+        // down why results doesn't come up (for example in situation when theres bad analyzer that takes ages to complete).
+        private void LogTimeouts(Task task, string description)
+        {
+            if (!task.IsCanceled) _logger.LogError($"Timeout before work got ready for {description}.");
         }
 
-        private async Task Dequeue()
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs changeEvent)
```

Comment:
Roslyn has a a [lot of tests](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.UnitTests/SolutionCrawler/WorkCoordinatorTests.cs,36) for their equivalent feature. I'm having trouble understanding what the exact behavior of this class is supposed to be and could probably easily regress it with the existing test coverage. Can we add more granular tests for what's supposed to happen?

Reply:
Most of complexity of service comes from part that consumer (at least vscode) excepts that when it requests current state of diagnostics it actually gets current state but analysis takes while and will be eventually ready. In same time end user might be writing code that causes even more stuff to be added but api should return last correct diagnostic results for current pending calls. Without additional logic diagnostics were constantly lagging behind.

I think correct solution is to move this painful part out from diagnostic service itself and test it with unit tests. This because testing behavior like this with host level tests are not really applicaple with missing control to asyncronity and timing. `PollingDiagnosticApiAdapter.cs` that contains this and is tested with separatly with unit tests.

However should these be implemented before or after this PR? There are real possibility that this issue fixes itself when file based api is used magically. But i don't know it yet what it will look like. It will be near end of year once i at least have time to dig that dept. I am not fan of untested code either but this feature will take long time to perfect 😞 

Workaround (retries) were added to tests because it seems sometimes order of event about workspace updates are invoked after results are fetched from API during tests. However this requires very specific timing which i couldn't repro with own machine 🤔 There was mention somewhere which i didn't now find that there is similar issue with LSP version sometimes where those events were missed, but couldn't find it now. Dunno are those two things connected however.

Reply:
What's the "file-based API"? You mentioned it in a couple of places and I'm not sure what you mean.

If trying to shim the existing behavior with the new implementation is this hard, we can also introduce a new version of the /diagnostics and /codecheck endpoints, say "v3" that have different semantics. As you described I think the existing implementation assumes it's going to get diagnostics for every file in the world when the request is complete. That doesn't really match the behavior the new system, which incrementally adds diagnostics. If we could version the API and change C# extension and OmniSharp, what would the interaction look like?

I feel that a PR of this scope needs to include comprehensive tests. Tests make it:
* Easier for us to understand what the code does
* Possible to maintain/update the behavior without breaking things

Thoughts on ^^^ @filipw @david-driscoll @DustinCampbell ?

Reply:
"file-based api" -> document based roslyn api (mistake), `await document.GetSemanticModelAsync()` etc. Analyzers currently use project based roslyn apis, which basically means code like:

```
var compiled = await project.WithCompilationOptions(
                    _rulesetsForProjects.BuildCompilationOptionsWithCurrentRules(project))
                    .GetCompilationAsync(token);

await compiled
                        .WithAnalyzers(allAnalyzers, workspaceAnalyzerOptions)
                        .GetAllDiagnosticsAsync(token);
```

Terminology might be wrong, this is first time i am working with roslyn at this level.

However as i was told there should be similar api to documents which i don't yet know how it is invoked. At that point analysis can be made document basis as it was before.

> If trying to shim the existing behavior with the new implementation is this hard

Thats basically whats most parts diagnostic service is doing now. Tokens in dictionaries etc are for that one and really have nothing to do with analysis itself. They can be moved out and tested separatly. And yes it was surpisingly tricky to get working correctly in both TestHost level tests and real world, there was much simpler versions during the way.

Reply:
I think i got idea how shim is easy to make separate / testable logic out from analyzer, i will write it this weekend.

Reply:
> What's the "file-based API"? You mentioned it in a couple of places and I'm not sure what you mean.

I would imagine this is the API in question. http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationWithAnalyzers.cs,562 as it allows you to ask for semantic diagnostics for a given semantic model only. There is a corresponding one for syntax diagnostics http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationWithAnalyzers.cs,486. This way it could be scoped to a single document.